### PR TITLE
Add crate-level documentation

### PR DESCRIPTION
**Pull request by Nashtare**
Add crate-level documentation and API explanations for the evm crate.

Should address #1376

### CODE REVIEW

**Review by muursh — APPROVED** [review]
I'll hold back judgement as to whether it meets to desires of 1367 but what is here looks fine to me.

**Comment by Nashtare** [comment]
I guess @pgebheim is the best to answer this question!

File: evm/src/lib.rs
Comment (dvdplm) [inline evm/src/lib.rs:81]:
```suggestion
//!     prover_state.prove_root(all_stark, config, inputs_3, &mut timing, None);
```

File: evm/src/lib.rs
Comment (dvdplm) [inline evm/src/lib.rs:93]: This was a question I had: are we guaranteed to fail if/when states are not contiguous (but otherwise valid)?

File: evm/src/lib.rs
Comment (dvdplm) [inline evm/src/lib.rs:60]: If the PVs are included in the proof, why are they so often passed around separately? Is it expensive to recover them from the proof's encoded public values?

File: evm/src/lib.rs
Comment (dvdplm) [inline evm/src/lib.rs:112]: > undesirable issues

Such as?

File: evm/src/lib.rs
Comment (dvdplm) [inline evm/src/lib.rs:133]: Can you provide an order of magnitude for how big the proofs are when serialized?

File: plonky2/src/fri/mod.rs
Comment (dvdplm) [inline plonky2/src/fri/mod.rs:18]: Should we provide a sane `Default` impl for this you think?

File: evm/src/config.rs
Comment (Nashtare) [reply evm/src/config.rs:5]: That's a good suggestion, this is already done for `CircuitConfig` with the default `standard_recursion_config()`.

File: evm/src/fixed_recursive_verifier.rs
Comment (Nashtare) [reply evm/src/fixed_recursive_verifier.rs:408]: I was puzzled as well when writing it, would this work better?

```rust
/// - `degree_bits_ranges`: the logarithmic ranges to be supported for the recursive tables.
/// Transactions may yield arbitrary trace lengths for each STARK module (within some bounds),
/// unknown prior generating the witness to create a proof. Thus, for each STARK module, we
/// construct a map from `2^{degree_bits} = length` to a chain of shrinking recursion circuits,
/// starting from that length, for each `degree_bits` in the range specified for this STARK module.
/// Specifying a wide enough range allows a prover to cover all possible scenarios.
```

File: evm/src/lib.rs
Comment (Nashtare) [reply evm/src/lib.rs:93]: Yes, because we try connecting the states (with some equality checks in the circuit), for instance that the gas used after txn 1 is equal to the gas used prior txn 2. So the prover would not be able to generate a proof.

File: evm/src/lib.rs
Comment (Nashtare) [reply evm/src/lib.rs:60]: It's more as a matter of convenience than anything else. It shouldn't be too heavy to retrieve them from the encoded public inputs, but as they are there anyway, it cost nothing to attach them to the output proof. I guess to be a tad cleaner, we could have a custom proof type, like `EvmProof` that consists in a `ProofWithPublicInputs` and `PublicValues`, to hide the tuple behind.

File: evm/src/lib.rs
Comment (Nashtare) [reply evm/src/lib.rs:133]: the proofs or the prover state?

File: evm/src/lib.rs
Comment (Nashtare) [reply evm/src/lib.rs:112]: If proving is decentralized for instance, and a prover falls behind. Or if a prover was to crash and lose its DB, with previously generated proofs, it would need to start again from genesis.

File: plonky2/src/fri/mod.rs
Comment (Nashtare) [reply plonky2/src/fri/mod.rs:18]: This one may be less relevant, because typically `StarkConfig` and `CircuitConfig` may deserve their own `Default` impl, but they have drastically different `FriConfig`, one tailored for proving time, the other one for proof size and recursion efficiency. So I'd rather not here.

File: evm/src/lib.rs
Comment (Nashtare) [reply evm/src/lib.rs]: The prover state will range between 20GB to 25GB when serialized.
The proofs are in the ballpark of ~150kB, regardless of the aggregation level, based on the current config (although the initial STARK proofs for a transaction, depend on the actual transaction execution, and may be several MBs large).

File: plonky2/src/fri/mod.rs
Comment (dvdplm) [reply plonky2/src/fri/mod.rs:18]: Fair, your call. The counter argument would be: if incompetent people try to use some code, then we should help them out by providing a default that is at least safe-ish so they don't accidentally run with unsafe parameters (I don't know if this can happen here).

File: evm/src/lib.rs
Comment (dvdplm) [reply evm/src/lib.rs]: prover state! I'm picturing some poor devops person that is trying to figure out how to set up a proving machine. Could be useful to have a hunch of how much disk space is actually needed.

File: evm/src/lib.rs
Comment (dvdplm) [reply evm/src/lib.rs:112]: decentralized? You mean distributed yeah?

So by "undesirable issues" you mean "avoiding restarting from scratch after a crash or missing previous data"? Might as well write that then! :)

Do you think we should mention that using the checkpoint feature is probably not a good idea for production systems (but awesome for development)?

File: plonky2/src/fri/mod.rs
Comment (Nashtare) [reply plonky2/src/fri/mod.rs:18]: We can't really have unsafe parameters here, because if for some reasons those were unsound (typically not matching the targeted security level), one could not create a `CircuitBuilder` instance, needed to construct circuits (it checks it).

The only risk is that someone may use a configuration not suited for their application, but well in general I wouldn't recommend people not familiar at all with this kind of things to even think using such library for anything else than fun/learning purposes 🙃

File: evm/src/lib.rs
Comment (dvdplm) [reply evm/src/lib.rs:60]: It's aesthetically a bit annoying to pass around duplicate data, but if it's convenient as-is I'll shut up.

File: evm/src/lib.rs
Comment (Nashtare) [reply evm/src/lib.rs]: The alternative is to split the prover state (it's been rendered possible with #1438), by loading only upper circuits needed after shrinking the STARK proofs and split the transaction proof process into three chunks:

- compute the initial STARK proofs
- load the related recursive table circuits
- finish the txn proof (basically what's in `prove_root` after the first line)

But this is out of scope of the plonky2_evm lib.

File: evm/src/fixed_recursive_verifier.rs
Comment (dvdplm) [reply evm/src/fixed_recursive_verifier.rs:408]: Much better! 🙏

**Review by dvdplm — APPROVED** [review]
(Would still be good to get a review from @4l0n50 before merging)

File: evm/src/lib.rs
Comment (Nashtare) [reply evm/src/lib.rs:112]: I did mean decentralized. We could picture a rewarding scenario where first prover to submit a valid proof from checkpoint A to checkpoint B is rewarded.

Even in the case of centralized proving, the checkpoint feature is useful for the other issues I mentioned.

Comment (Nashtare) [reply evm/src/lib.rs:112]: Checkpointing can really be seen as skipping unnecessary work. Once you've submitted your proof on your settlement layer, and it's been verified and approved, your state up to block N is verified. Hence no need to keep track to what previously happen, and locally do a proof from block N to block N+c for some constant c. Proper connection between the two proofs should be done on the settlement layer, to ensure that the starting state of the incoming proof was the final state of the previously submitted proof (much like what's being done when aggregating transactions, except that this is done internally in the circuits).

File: evm/src/lib.rs
Comment (4l0n50) [inline evm/src/lib.rs:111]:
```suggestion
//! The process of always providing a previous block proof when generating a proof for the current block may yield some
```

File: evm/src/lib.rs
Comment (4l0n50) [inline evm/src/lib.rs:112]:
```suggestion
//! undesirable issues. For this reason the plonky2 zk-EVM supports checkpoint heights. At given block heights, the prover does not have to pass
```

Comment (4l0n50) [reply evm/src/lib.rs:112]: It's just because the "Because of..." is also used later

**Review by 4l0n50 — COMMENTED** [review]
Looks good to me. Just minor comments

File: evm/src/lib.rs
Comment (Nashtare) [reply evm/src/lib.rs:92]: Typo on my side, it's actually `prove_aggregation`! Fixed it, thanks!

**Pull request merged**
